### PR TITLE
Imu driver

### PR DESCRIPTION
**Pull request by HamzaIqbal69**
Filled in IMU driver

### CODE REVIEW

File: Drivers/imu/imu.c [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    LSM6DSOXTR IMU DRIVER Source File
+    Link to part Datasheet for reference:
+    https://www.st.com/resource/en/datasheet/lsm6dsox.pdf
+
+    Author: Hamza Iqbal
+*/
+#include "imu.h"
+
+/* IMU Intitialization */
+uint8_t imu_init(IMU_DATA *imu, I2C_HandleTypeDef *i2cHandle)
```
Comment (nwdepatie) [inline Drivers/imu/imu.c:11]: I think you forgot to rename some of your parameters, since it still says IMU_DATA. Also do i2c_handle

**Review by nwdepatie — CHANGES_REQUESTED** [review]
Overall I am really impressed. I would merge it in now but there's some minor fixes I want you to think about. I think one more revision and we can squash and merge

File: Drivers/imu/imu.h [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+#define IMU_DRIVER_H
+#include "stm32f4xx_hal.h"
+
+#define IMU_I2C_ADDRESS                 0x35 << 1   /* Shifted because datasheet said to */
+// Not sure if these are all needed, also not sure if more need ot be added
+/* For register descriptions reference datasheet pages 47 - 98 */
+#define LSM6D_REG_FUNC_CFG_ACCESS       0x01    /* Enable embedded functions register */
+#define LSM6D_REG_INTERRUPT_CTRL_1      0x0D    /* INT1 pin control, used for interrupts */
+#define LSM6D_REG_INTERRUPT_CTRL_2      0x0E    /* INT2 pin control, used for interrupts */
+#define LSM6D_REG_WHO_AM_I              0x0F    /* Register for checking communication */
```
[the remark this answers is not in the document]
Comment (HamzaIqbal69) [reply Drivers/imu/imu.h:19]: Thats the actual name but i can change that

File: Drivers/imu/imu.c [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+    uint8_t error_num = 0;
+    HAL_StatusTypeDef status;
+
+    /* Quick check to make sure I2C is working */
+    uint8_t* regData;
+    status = imu_read_reg(imu, regData, LSM6D_REG_WHO_AM_I);
+    err_num += (status != HAL_OK);
+    if(*regData != 0x6C)
+    {
+        return 255;
```
[the remark this answers is not in the document]
Comment (HamzaIqbal69) [reply Drivers/imu/imu.c:33]: I just used that arbitrarily as a fault code, I could enum it

File: Drivers/imu/imu.c [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+    }
+    
+    /* Configure IMU Refer to datasheet page 56 - 57 */    
+
+    /* Accelerometer configuration settings */
+    int8_t odr_sel_accel = 0x08;
+    int8_t fs_sel_accel = 0x02;
+    int8_t lp_f2_enable_accel = 0x01;
```
[the remark this answers is not in the document]
Comment (HamzaIqbal69) [reply Drivers/imu/imu.c:41]: not sure how to make this work since I need it to be an 8 bit integer

File: Drivers/imu/imu.c [hunk 5]
```diff
@@ -0,0 +1,171 @@
...
+    LSM6DSOXTR IMU DRIVER Source File
+    Link to part Datasheet for reference:
+    https://www.st.com/resource/en/datasheet/lsm6dsox.pdf
+
+    Author: Hamza Iqbal
+*/
+#include "imu.h"
+
+/* IMU Intitialization */
+uint8_t LSM6D_init(IMU_DATA *imu, I2C_HandleTypeDef *i2cHandle)
```
Comment (nwdepatie) [inline Drivers/imu/imu.c:11]: Still an IMU_DATA :(

**Review by nwdepatie — APPROVED** [review]
I'm approving since I made the changes I wanted lol

**Comment by nwdepatie** [comment]
I was lazy and forgot to separate the commits for generating the STM32 HAL code and writing the driver, in the future I'll try and separate those into different commits

**Pull request merged**